### PR TITLE
types,docker: add DockerAuthConfig

### PR DESCRIPTION
**Pull request by runcom**
The automatic credentials retrieval built into containers/image doesn't play at all with external credentials store. Furthermore, we need a way to pass credentials to containers/image from docker/docker (projectatomic/docker#200). This also fixes a bunch of docker's integration tests in projectatomic/docker#200.

This patch just gives precedence to any `ctx.DockerAuthConfig` field before diving deep into the system to lookup credentials into `~/.docker/config.json`. So no real behavior change in skopeo so far. We'll need to tackle #111, at which point we'll made the automatic retrieval optional as per https://github.com/containers/image/pull/41#issuecomment-233338484.

Close #41 
Close projectatomic/skopeo#106

This is also needed by kubernetes-incubator/cri-o#8 since there it's k8s itself to provide those credentials and we need to honor that (instead of looking on the system).

@mtrmac PTAL this is blocking projectatomic/docker#200 as said also. 

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

### CODE REVIEW

**Comment by runcom** [comment]
LGTM

File: types/types.go [hunk 1]
```diff
@@ -223,4 +240,7 @@ type SystemContext struct {
 	// === docker.Transport overrides ===
 	DockerCertPath              string // If not "", a directory containing "cert.pem" and "key.pem" used when talking to a Docker Registry
 	DockerInsecureSkipTLSVerify bool   // Allow contacting docker registries over HTTP, or HTTPS with failed TLS verification. Note that this does not affect other TLS connections.
+	// if nil, the library tries to parse ~/.docker/config.json to retrieve credentials
+	// FIXME: the automatic credentials retrieval doesn't play with external stores also, see containers/image#111
```
Comment (mtrmac) [inline types/types.go]: This is unrelated to this field AFAICS; if we are already tracking this as an issue, no need to add a FIXME here.

File: docker/docker_client.go [hunk 2]
```diff
@@ -243,7 +243,13 @@ func (c *dockerClient) getBearerToken(realm, service, scope string) (string, err
 	return tokenStruct.Token, nil
 }
 
-func getAuth(hostname string) (string, string, error) {
+func getAuth(ctx *types.SystemContext, hostname string) (string, string, error) {
+	if ctx.DockerAuthConfig != nil {
+		if ctx.DockerAuthConfig.Auth != "" {
+			return decodeDockerAuth(ctx.DockerAuthConfig.Auth)
+		}
+		return ctx.DockerAuthConfig.Username, ctx.DockerAuthConfig.Password, nil
```
Comment (mtrmac) [inline docker/docker_client.go]: Why support both `Auth` and `Username/Password` here? Do we already have real users for both which would benefit? Then we need to document which overrides what, it just seems like unnecessary API complexity.

I’d like to have only the more explicit and type-safe username/password pair, if there were no other reason to prefer the `Auth` one or to supply both.

Comment (mtrmac) [inline docker/docker_client.go]: (It is not _completely_ certain that we can always ignore the `hostname` here, but it is likely enough to be the common case; we can always add a `DockerAuthConfigMap` later, so this is fine.)

File: types/types.go [hunk 3]
```diff
@@ -200,6 +200,23 @@ type ImageInspectInfo struct {
...
+	Auth     string
+	// FIXME: currently not used
+	ServerAddress string
+	// IdentityToken is used to authenticate the user and get
+	// an access token for the registry.
+	// FIXME: currently not used
+	IdentityToken string
+	// RegistryToken is a bearer token to be sent to a registry
+	// FIXME: currently not used
+	RegistryToken string
```
Comment (mtrmac) [inline types/types.go]: I guess, drop the unused fields if we don’t know how we would be using them?

Especially the `ServerAddress` makes no sense to me considering the way `getAuth` is called.

File: types/types.go [hunk 4]
```diff
@@ -200,6 +200,23 @@ type ImageInspectInfo struct {
 	Layers        []string
 }
 
+// DockerAuthConfig contains authorization information for connecting to a registry.
+// FIXME: forked from docker/docker, waiting a full docker re-vendor
+type DockerAuthConfig struct {
+	Username string
+	Password string
+	Auth     string
```
Comment (mtrmac) [inline types/types.go]: `Auth`, if we have it at all (see elsewhere) should be documented.

File: types/types.go (hunk 3 above)
Comment (mtrmac) [inline types/types.go]: (Do we even need an extra struct, or just more members in `SystemContext`?

File: docker/docker_client.go [hunk 5]
```diff
@@ -243,7 +243,13 @@ func (c *dockerClient) getBearerToken(realm, service, scope string) (string, err
 	return tokenStruct.Token, nil
 }
 
-func getAuth(hostname string) (string, string, error) {
+func getAuth(ctx *types.SystemContext, hostname string) (string, string, error) {
+	if ctx.DockerAuthConfig != nil {
```
Comment (mtrmac) [inline docker/docker_client.go]: This needs a `ctx != nil && `

File: types/types.go (hunk 3 above)
Comment (runcom) [inline types/types.go]: Ocid is going to be an user of this whole struct as soon as containers/image is plumbed there.. This is somehow copied from that..

File: docker/docker_client.go (hunk 2 above)
Comment (runcom) [inline docker/docker_client.go]: Ocid can provide username/password from k8s or skopeo can. K8s and OpenShift can just provide Auth.

File: types/types.go (hunk 3 above)
Comment (mtrmac) [inline types/types.go]: Then we need to also implement all of that?

File: docker/docker_client.go (hunk 2 above)
Comment (mtrmac) [inline docker/docker_client.go]: > K8s and OpenShift can just provide Auth. 

AFAICT for access to the Docker endpoint for OpenShift, one would set `{Username:"unused",Password:openShiftTokenString}`, at least based on what `~/.kube/config` is storing. Does the API really give the user the Docker base64 blob somewhere?

Comment (runcom) [inline docker/docker_client.go]: Alright, not sure about origin then, but that struct is exactly the same in k8s so it might happen k8s sends it that way

File: types/types.go (hunk 3 above)
Comment (runcom) [inline types/types.go]: I guess so, but we're far from fully doing that so this whole PR is just unblocking projectatomic/docker#200 while adding this struct witch is the same in docker and k8s api

File: types/types.go (hunk 1 above)
Comment (runcom) [inline types/types.go]: ack

File: types/types.go (hunk 3 above)
Comment (mtrmac) [inline types/types.go]: So, grepping `docker/docker`:
- `RegistryToken` is set nowhere in the code (but could be provided by a third-party client using the Remote API)
- `IdentityToken` is `refresh_token` from https://github.com/docker/distribution/blob/master/docs/spec/auth/token.md , something we neither use nor generate (but we could, I suppose)
- `ServerAddress` seems to be a data structure artifact (look how `cliconfig/configfile/file.go:LoadFromReader` overrides the value in the unmarshaled struct) and/or a semantic mix-up (`registry/service.go` uses `AuthConfig.ServerAddress` to look up registry hostnames)
- `Auth` also an artifact of reusing the same structure for internal data and config storage, it is not really used anywhere, except `cliconfig/configfile/file.go`, which decodes the field _and zeroes it_.

Grepping `openshift/origin`:
- No use of `ReghistryToken`, `IdentityToken`
- AFAICS from a _very_ imprecise grep the only real use is in `vendor/k8s.io/kubernetes/pkg/credentialprovider/keyring.go` , which only sets `Username`, `Password`, `Email`; I might have missed something.

**Comment by runcom** [comment]
@mtrmac not sure where we're on this one. Would you have anything changed? I'd love to keep that struct as is right now since it reflects what's in k8s and since we didn't make any API stability promise we can always change it later.

**Comment by mtrmac** [comment]
I’d rather have a struct which contains _only_ the fields we are implementing; and to add other fields later if and when we implement them, and only those which make sense. That way, we actually benefit from our ability to change the API later, not by adding fields just in case we might possibly use some of them in the future.

**Comment by runcom** [comment]
> I’d rather have a struct which contains only the fields we are implementing; and to add other fields later if and when we implement them, and only those which make sense. That way, we actually benefit from our ability to change the API later, not by adding fields just in case we might possibly use some of them in the future.

ack, User, Password and Auth right? I'll remove the others

**Comment by mtrmac** [comment]
Not `Auth`; `Auth` is set to `""` just after parsing the config.

… actually, one thing I forgot: apparently the Docker Remote API allows sending an arbitrary struct (see the `RegistryToken` comment); so it’s not obvious that it is tenable for Docker to just ignore some fields. I suppose _something_ should fail if a field is unimplemented and ignored; whether that is docker/docker or this package then doesn’t matter that much.

**Comment by runcom** [comment]
Shall we leave that to docker/docker then?

**Comment by mtrmac** [comment]
I think so, that’s the API server. Unless k8s happens to do exactly the same thing which would be shared?

(Another minor point: If we are declaring a new `struct` in `containers/image/types`, the original `docker/engine-api/types.AuthConfig` is not directly assignable to ours anyway, is it? It needs to be a filed-by-field copy in any case.)

**Comment by runcom** [comment]
> (Another minor point: If we are declaring a new struct in containers/image/types, the original docker/engine-api/types.AuthConfig is not directly assignable to ours anyway, is it? It needs to be a filed-by-field copy in any case.)

right, I will take care of that.

**Comment by runcom** [comment]
@mtrmac PTAL

**Comment by mtrmac** [comment]
Yeah, but tests are still broken after #115.

**Comment by runcom** [comment]
Fixing those

**Comment by runcom** [comment]
> Yeah, but tests are still broken after #115.

tests fixed over there https://github.com/projectatomic/skopeo/pull/226

**Comment by runcom** [comment]
I'll make a skopeo PR for this change after I merge #115. Then I'll merge this one if you LGTM this as is (as I think you did?)

**Comment by runcom** [comment]
I've restarted the travis build here meanwhile..

**Comment by runcom** [comment]
tests green here

**Comment by mtrmac** [comment]
:+1:
